File: ts/findMinArrowShots.ts
function findMinArrowShots(points: number[][]): number {
  if (points.length === 0) return 0;
  points.sort((a, b) => a[1] - b[1]);

  let count = 1;
  let [_, max] = points[0];

  for (let i = 1; i < points.length; i++) {
    const [start, end] = points[i];

    if (start <= max) {
      continue;
    }

    max = end;
    count += 1;
  }

  return count;
}

export default () => {
  const points = [
    [1, 2],
    [2, 3],
    [3, 4],
    [4, 5],
  ];
  // const points = [
  //   [1, 2],
  //   [3, 4],
  //   [5, 6],
  //   [7, 8],
  // ];
  // const points = [
  //   [10, 16],
  //   [2, 8],
  //   [1, 6],
  //   [7, 12],
  // ];
  const result = findMinArrowShots(points);
  console.log(result);
};

File: rs/src/find_min_arrow_shots.rs
fn find_min_arrow_shots(mut points: Vec<Vec<i32>>) -> i32 {
    if points.len() == 0 {
        return 0;
    }
    points.sort_by(|a, b| a[1].cmp(&b[1]));

    let mut count: i32 = 1;
    let mut max = points[0][1];

    for i in 1..points.len() {
        if points[i][0] <= max {
            continue;
        }

        max = points[i][1];
        count += 1;
    }

    return count;
}

pub fn main() {
    let points: Vec<Vec<i32>> = vec![vec![10, 16], vec![2, 8], vec![1, 6], vec![7, 12]];
    let result = find_min_arrow_shots(points);

    println!("{result}");
}
